### PR TITLE
X25519 encryption

### PR DESCRIPTION
**Pull request by ccorcoveanu**

### CODE REVIEW

File: encryption/x25519/scalar.go [hunk 1]
```diff
@@ -0,0 +1,212 @@
+package x25519
+
+import (
+	"bytes"
+	"crypto/sha512"
+	"github.com/ElrondNetwork/elrond-go-core/core/check"
+	crypto "github.com/ElrondNetwork/elrond-go-crypto"
+	"golang.org/x/crypto/curve25519"
+)
+
```
Comment (AdoAdoAdo) [inline encryption/x25519/scalar.go:10]: can you add a 
```go
var _ crypto.Scalar = (*x25519Scalar)(nil)
```

File: encryption/x25519/pubKeyEncryptor.go [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+	if err != nil {
+		return nil, err
+	}
+	signer := singlesig.Ed25519Signer{}
+	err = signer.Verify(originatorPubKey, authMessage.Sum(nil), macBytes)
+	if err != nil {
+		return nil, err
+	}
+
+	var recipientX25519 x25519Point
```
Comment (AdoAdoAdo) [inline encryption/x25519/pubKeyEncryptor.go:144]: can you extract the decryption in a private decrypt function and call it here?

**Review by AdoAdoAdo — CHANGES_REQUESTED** [review]

File: encryption/x25519/point.go [hunk 3]
```diff
@@ -0,0 +1,139 @@
+package x25519
+
+import (
+	"bytes"
+	"filippo.io/edwards25519"
+	"github.com/ElrondNetwork/elrond-go-core/core/check"
+	crypto "github.com/ElrondNetwork/elrond-go-crypto"
+)
+
```
[the remark this answers is not in the document]
Comment (ccorcoveanu) [reply encryption/x25519/point.go:9]: done

File: encryption/x25519/suite.go [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+
+// ScalarLen returns the length of the x25519 private key
+func (s *suiteX25519) ScalarLen() int {
+	return curve25519.ScalarSize
+}
+
+// CreateScalar creates a new Scalar which represents the x25519 private key
+func (s *suiteX25519) CreateScalar() crypto.Scalar {
+	sk := make([]byte, curve25519.ScalarSize)
+	if _, err := rand.Read(sk); err != nil {
```
[the remark this answers is not in the document]
Comment (ccorcoveanu) [reply encryption/x25519/suite.go:78]: was an import mistake, fixed, good catch 👍

File: go.mod [hunk 5]
```diff
@@ -3,9 +3,10 @@ module github.com/ElrondNetwork/elrond-go-crypto
 go 1.13
 
 require (
+	filippo.io/edwards25519 v1.0.0
 	github.com/ElrondNetwork/elrond-go-core v1.0.0
 	github.com/ElrondNetwork/elrond-go-logger v1.0.5
 	github.com/herumi/bls-go-binary v1.0.0
 	github.com/stretchr/testify v1.7.0
-	golang.org/x/crypto v0.0.0-20210711020723-a769d52b0f97
+	golang.org/x/crypto v0.1.0
```
[the remark this answers is not in the document]
Comment (ccorcoveanu) [reply go.mod]: I did not see the tag on github, updated

File: encryption/x25519/pubKeyEncryptor.go [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+	originatorPubKeyBytes, err := hex.DecodeString(ed.Identities.OriginatorPubKey)
+	if err != nil {
+		return nil, err
+	}
+	macBytes, err := hex.DecodeString(ed.Crypto.MAC)
+	if err != nil {
+		return nil, err
+	}
+
+	// Verification of the auth message
```
[the remark this answers is not in the document]
Comment (ccorcoveanu) [reply encryption/x25519/pubKeyEncryptor.go:124]: done

File: encryption/x25519/pubKeyEncryptor.go (hunk 2 above)
Comment (ccorcoveanu) [reply encryption/x25519/pubKeyEncryptor.go:144]: done

**Review by AdoAdoAdo — DISMISSED** [review]

File: encryption/x25519/scalar.go [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+// SetInt64 is not needed for this use case, should be removed if possible
+func (x *x25519Scalar) SetInt64(_ int64) {
+	log.Error("x25519Scalar",
+		"message", "SetInt64 for x25519Scalar is not implemented, should not be called")
+}
+
+// Zero is not needed for this use case, should be removed if possible
+func (x *x25519Scalar) Zero() crypto.Scalar {
+	log.Error("x25519Scalar",
+		"message", "SetInt64 for x25519Scalar is not implemented, should not be called")
```
Comment (sstanculeanu) [inline encryption/x25519/scalar.go:112]:
```suggestion
		"message", "Zero for x25519Scalar is not implemented, should not be called")
```

File: encryption/x25519/pubKeyEncryptor.go [hunk 8]
```diff
@@ -0,0 +1,261 @@
...
+//  for Alice (remember that the private key used for encryption
+//  is ephemeral - in order to avoid nonce reuses and
+//  multipurpose use of the same secret)
+type EncryptedDataIdentities struct {
+	Recipient        string `json:"recipient"`
+	EphemeralPubKey  string `json:"ephemeralPubKey"`
+	OriginatorPubKey string `json:"originatorPubKey"`
+}
+
+type EncryptedCryptoData struct {
```
Comment (sstanculeanu) [inline encryption/x25519/pubKeyEncryptor.go:33]: missing comment

File: encryption/x25519/scalar.go [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+	"github.com/ElrondNetwork/elrond-go-core/core/check"
+	crypto "github.com/ElrondNetwork/elrond-go-crypto"
+	"golang.org/x/crypto/curve25519"
+)
+
+var _ crypto.Scalar = (*x25519Scalar)(nil)
+
+type PrivateKey []byte
+
+func (p *PrivateKey) Public() (PublicKey, error) {
```
Comment (sstanculeanu) [inline encryption/x25519/scalar.go:15]: missing comment

File: encryption/x25519/suite.go [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+	crypto "github.com/ElrondNetwork/elrond-go-crypto"
+	"golang.org/x/crypto/curve25519"
+)
+
+// X25519 is the string representations of the X25519 scheme
+const X25519 = "X25519"
+
+type suiteX25519 struct{}
+
+func NewX25519() *suiteX25519 {
```
Comment (sstanculeanu) [inline encryption/x25519/suite.go:16]: missing comment

File: encryption/x25519/suite.go [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+}
+
+// CheckPointValid -
+func (s *suiteX25519) CheckPointValid(pointBytes []byte) error {
+	if len(pointBytes) != s.PointLen() {
+		return crypto.ErrInvalidParam
+	}
+
+	point := s.CreatePoint()
+	err := point.UnmarshalBinary(pointBytes)
```
Comment (sstanculeanu) [inline encryption/x25519/suite.go:105]:
```suggestion
	return point.UnmarshalBinary(pointBytes)
```

File: encryption/x25519/pubKeyEncryptor.go (hunk 8 above)
Comment (ccorcoveanu) [reply encryption/x25519/pubKeyEncryptor.go:33]: done

File: encryption/x25519/scalar.go (hunk 9 above)
Comment (ccorcoveanu) [reply encryption/x25519/scalar.go:15]: done

File: encryption/x25519/suite.go (hunk 10 above)
Comment (ccorcoveanu) [reply encryption/x25519/suite.go:16]: done

File: encryption/x25519/suite.go [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+
+// GetUnderlyingSuite returns nothing because this is not a wrapper over another suite implementation
+func (s *suiteX25519) GetUnderlyingSuite() interface{} {
+	log.Warn("suiteX25519",
+		"message", "calling GetUnderlyingSuite for suiteX25519 which has no underlying suite")
+
+	return nil
+}
+
+// CheckPointValid -
```
[the remark this answers is not in the document]
Comment (ccorcoveanu) [reply encryption/x25519/suite.go:98]: done

**Review by sstanculeanu — DISMISSED** [review]
should base branch be updated to rc/v1.5.0?

**Review by AdoAdoAdo — APPROVED** [review]

**Review by AdoAdoAdo — APPROVED** [review]

**Review by sstanculeanu — APPROVED** [review]

**Pull request merged**